### ISSUE TITLE
Possible rustc language features for less magic?

### ISSUE THREAD

**Issue by johannescpk** [idanarye/rust-typed-builder#71]
Was just thinking which kind of language features could help with a cleaner approach compared "impl on tuple shadowing" and "sealed enums" – which is a creative way to solve it for sure :+1:. Are you already aware of RFCs that touch on those? Something that came to mind, but no idea if it's feasable: "`Option` must be `Some`"-attribute that makes the compiler panic if option is `None`

```rust
struct UwuBuilder {
  #[option_must_be_some_panic(message = "owo is required")]
  owo: Option<usize>,
}
```

**Comment by idanarye** [comment]
If an `Option` cannot be `None` - why not just make it a non-`Option`?

**Comment by johannescpk** [comment]
```rust
struct UwuBuilder {
  /// required
  #[option_must_be_some_compile_error(message = "owo is required")]
  owo: Option<usize>,

  /// optional
  uwu: Option<usize>,
}

impl UwuBuilder {
  fn build() {}
}

UwuBuilder { owo: Some(1), uwu: None }.build(); // compiles
UwuBuilder { owo: None, uwu: None }.build(); // Does not compile
```

That would be idea. Makes more sense?

**Comment by idanarye** [comment]
Why not this?

```rust
#[derive(Default)]
struct UwuBuilder {
  /// required
  owo: usize,

  /// optional
  uwu: Option<usize>,
}
```

**Comment by johannescpk** [comment]
Then you can't have a builder, can you?

**Comment by idanarye** [comment]
Oh. I see your point now - I've somehow missed that you were defining an `UwuBuilder` struct and not an `Uwu` struct. But in this crate I went for a different approach - you would define the `Uwu` struct and let the derive macro create the `UwuBuilder`, which is generic and its generic parameter encodes which fields were set already. So an `UwuBuilder` where `owo` is set has a different type (because of the generic parameter) than an `UwuBuilder` where `owo` is not set, and only the former will have a `build` method.

**Comment by johannescpk** [comment]
Yeah, I know how your crate and is working. The issue was meant to talk about possible language features to not require the hacks that your crate is doing to enable the behavior it's providing – because generating impls on different combinations of tuples and having an enum without variants so you can't call the build function are hacks. Creative hacks, but hacks.

**Comment by johannescpk** [comment]
Sorry if that sounded a bit rude. And if you feel like this is out of scope for your crate specifically, feel free to close.

**Comment by idanarye** [comment]
Well, if we are talking about language features, how about something like this?
```rust
struct UwuBuilder<const D: Definition<Self>>  {
    /// required
    owo: usize,

    /// optional
    #[omittable]
    uwu: usize,
}

impl<const D: Definition<UwuBuilder>> UwuBuilder<D> {
    fn build(self) -> Uwu {
        Uwu {
            owo: self.owo,

            #[if(D.has("uwu"))]
            uwu: self.uwu,

            #[if(!D.has("uwu"))]
            uwu: 42,
        }
    }
}
```

And use it like this:
```rust
assert_eq(
    UwuBuilder {
        owo: 1,
    }.build(),
    Uwu {
        owo: 1,
        uwu: 42,
    },
);

assert_eq(
    UwuBuilder {
        owo: 2,
        uwu: 3,
    }.build(),
    Uwu {
        owo: 2,
        uwu: 3,
    },
);
```

**Comment by johannescpk** [comment]
Interesting approach!

But isn't the tricky part of the builder that you need to have all fields be `Option`s, so that the struct can get initialized completely empty, and then only fill the `Option`s with `Some` after constructing via method calls? So your

```
struct UwuBuilder<const D: Definition<Self>>  {
    /// required
    owo: usize,
```

wouldn't work, right?

**Comment by idanarye** [comment]
"Starting empty" is not really a core requirement of the builder pattern. One could imaging a builder that works like this:
```rust
assert_eq!(
    Uwu.builder(2).uwu(3).build(),
    Uwu { owo: 2, uwu: 3 },
);
```

Or even:
```rust
assert_eq!(
    UwuBuilder { owo: 2 }.uwu(3).build(),
    Uwu { owo: 2, uwu: 3 },
);
```
